### ISSUE TITLE
[Question] How to Weigh Certain Training Samples Differently Depending on the Label?

### ISSUE THREAD

**Issue by shoyasaxa** [NVIDIA-Merlin/HugeCTR#272]
Hello, 

I have both implicit and explicit labels that I'd like to weigh differently. Could you please provide guidance on how we could do that? 

I see that `BinaryCrossEntropy` only supports binary labels, but I also see that `CrossEntropyLoss` is implemented. 
However, I am a bit confused on how to use `CrossEntropyLoss`. In the document, it says that the input/output shapes are 

```
- input: [(batch_size, 2), (batch_size, 2)] where the first tensor represents the predictions while the second tensor represents the labels
- output: (batch_size, 2)
```

Why does the second dimension for the labels has to be `2`? 

I see from [here](https://github.com/NVIDIA-Merlin/HugeCTR/issues/144) that `CrossEntropyLoss` is supposed to mimic `tf.keras.losses.SparseCategoricalCrossentropy`, but from my understanding, `tf.keras.losses.SparseCategoricalCrossentropy`'s label tensor does not have to have a second dimension of `2`. TF's webpage [here](https://www.tensorflow.org/api_docs/python/tf/keras/losses/SparseCategoricalCrossentropy)  says it can do 
```
y_true = [1, 2]
y_pred = [[0.05, 0.95, 0], [0.1, 0.8, 0.1]]
# Using 'auto'/'sum_over_batch_size' reduction type.
scce = tf.keras.losses.SparseCategoricalCrossentropy()
scce(y_true, y_pred).numpy()
```

Thank you for your help!

**Comment by zehuanw** [comment]
Hi @shoyasaxa , thank you for your questions!

In `CrossEntropyLoss` the label will be either 1,0 or 0,1. That's why it needs dimension of 2. It's still a representation of `click or not`.

> I have both implicit and explicit labels that I'd like to weigh differently. Could you please provide guidance on how we could do that?
If you want Multilabel/Multitask loss,  we are going to support limited metrics like "AverageLoss" in the Dec release soon. Each label will have different weight then. But please be free to let us know if it's not what you want.

**Comment by shoyasaxa** [comment]
Hi @zehuanw, thank you for answering! That is great that multilabel/multitask loss is coming up! 

I just have one follow up question - if `CrossEntropyLoss` is still for binary labels, what is the difference between `CrossEntropyLoss` and `BinaryCrossEntropy` other than how the label is passed in?

**Comment by yingcanw** [comment]
> I just have one follow up question - if CrossEntropyLoss is still for binary labels, what is the difference between CrossEntropyLoss and BinaryCrossEntropy other than how the label is passed in?

@shoyasaxa Thanks for your feedback. As described in the [issue 144](https://github.com/NVIDIA-Merlin/HugeCTR/issues/144)，the implementation of  `CrossEntropyLoss`  in HugeCTR is equivalent to `tf.keras.losses.SparseCategoricalCrossentropy` . The definition of  is consistent with [tf2.1](https://tensorflow.google.cn/versions/r2.1/api_docs/python/tf/keras/losses/SparseCategoricalCrossentropy) instead of version [tf2.7](https://www.tensorflow.org/api_docs/python/tf/keras/losses/SparseCategoricalCrossentropy) you provided for your reference.

**Comment by yingcanw** [comment]
> I just have one follow up question - if CrossEntropyLoss is still for binary labels, what is the difference between CrossEntropyLoss and BinaryCrossEntropy other than how the label is passed in?
@shoyasaxa  Thanks for your feedback. You can understand that a final sigmoid function is fused with the loss function for `BinaryCrossEntropy` and a softmax function is fused with the loss function for `CrossEntropyLoss`.

**Comment by shoyasaxa** [comment]
Got it - thank you for the clarification!

**Comment by shoyasaxa** [comment]
Hi @zehuanw,

> If you want Multilabel/Multitask loss, we are going to support limited metrics like "AverageLoss" in the Dec release soon. Each label will have different weight then. But please be free to let us know if it's not what you want.

Has support for this been released yet? If so, could you point me to an example of how to accomplish that? Thank you!

**Comment by shoyasaxa** [comment]
Is it possible to use `MultiCrossEntropyLoss` with the `target_weight_vec` parameter in the final `hugectr.DenseLayer()` of a network to achieve weighing of different types of labels? 

If that is the case, I have a follow up question. Under instructions for `MultiCrossEntropyLoss` [here](https://github.com/NVIDIA-Merlin/HugeCTR/blob/v3.3/docs/hugectr_layer_book.md#multicrossentropyloss), it says
```
input: [(batch_size, *), (batch_size, *)] where the first tensor represents the predictions while the second tensor represents the labels. * represents any even number of elements.

Example:
model.add(hugectr.DenseLayer(layer_type = hugectr.Layer_t.MultiCrossEntropyLoss,
                            bottom_names = ["add", "label"],
                            top_names = ["loss"],
                            use_regularizer = True,
                            regularizer_type = hugectr.Regularizer_t.L1,
                            lambda = 0.1
                            ))
``` 
If we are using a parquet data format, what format does out `label` column need to be? Does it need to be an array column? For example, if there are 4 types of labels, does the `label` column need to be like 
```
___________
|  label   |
| [0,0,0,1]|
| [1,0,0,0]|
| [0,1,0,0]|
```
In addition, why does it say`* represents any even number of elements.`? In other words, why does the number of unique labels have to be an even number?

**Comment by yingcanw** [comment]
@shoyasaxa 

>  Is it possible to use MultiCrossEntropyLoss with the target_weight_vec parameter in the final hugectr.DenseLayer() of a network to achieve weighing of different types of labels?

You can use `MultiCrossEntropyLoss` as below:
```
model.add(hugectr.DenseLayer(layer_type = hugectr.Layer_t.MultiCrossEntropyLoss,
                            bottom_names = ["add", "label"],
                            top_names = ["loss"],
                            target_weight_vec = [0.1,0.9]))
```

> If we are using a parquet data format, what format does out label column need to be? Does it need to be an array column? For example, if there are 4 types of labels, does the label column need to be like

Please refer to the [Parquet Dataset Format](https://github.com/NVIDIA-Merlin/HugeCTR/blob/master/docs/python_interface.md#dataset-formats)

> In addition, why does it say* represents any even number of elements.? In other words, why does the number of unique labels have to be an even number?

It allows labels in an arbitrary dimension.

**Comment by shoyasaxa** [comment]
Got it - thank you!
